Swift 4.0 branch with exclusivity suppression

lhames opened the pull request:
Explanation:

These two commits teach LLDB to temporarily suppress enforcement of exclusivity during evaluation of swift expressions. This is necessary to allow debugger expressions to access variables that are exclusively owned in the original program in the context where the expression is to be evaluated.

Scope: Affects evaluation of swift expressions in LLDB. Will disable exclusivity enforcement by the swift runtime while each expression is executed, then re-enable it afterwards. This will suppress exclusivity errors in expressions, but should not affect expression evaluation otherwise.  

Radar (and possibly SR Issue): <rdar://problem/31715465>

Risk: Affects evaluation of swift expressions under LLDB only. Low risk.

Testing: Some testing by hand, plus test cases for:
(1) a single expression accessing an exclusively owned variable, and
(2) two nested expressions accessing an exclusively owned variable

lhames committed:
Add DidStartExecuting/WillFinishExecuting methods to Expression.

These methods can be used by the derived expression types to perform expression
specific and/or language specific actions before and after the expression runs.
(ThreadPlanCallUserExpression is modified to call these methods on the
expression immediately before/after execution of the expression).

The immediate motivation is allowing Swift expressions to notify the swift
runtime that exclusivity enforcement should be suspended while the expression
runs (we want LLDB expressions to be able to access variables even when they're
considered exclusively owned by someone else in the original program).

Reviewed in https://reviews.llvm.org/D32889

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@302314 91177308-0d34-0410-b5e6-96231b3b80d8

lhames committed:
Disable runtime enforcement of exclusivity during expression evaluation.

The debugger needs to be able to read the value of variables regardless of
whether they're marked as exclusively owned/locked by the swift runtime. This
patch modifies SwiftUserExpressions to poke the swift runtime before running
an expression to disable exclusivity enforcement, and again after the expression
has finished to re-enable enforcement.

<rdar://problem/31715465>

lhames commented:
@swift-ci please test

jimingham commented on packages/Python/lldbsuite/test/lang/swift/expression/exclusivity_suppression/TestExclusivitySuppression.py:
```diff
@@ -0,0 +1,145 @@
+        self.assertTrue(frame, "Frame 0 is valid.")
+
+        self.check_expression(frame, "w.s.i", "8", use_summary=False)
+
```
I don't think you finished this test.   It's pretty much identical to the first test (except for setting a second breakpoint.)

lhames replied:
Oops - this was meant to be left out of this commit and added in a follow-up.

I have a "complete" version of this test but it fails due to unexpected behavior in SBProcess::continue - I'll see if I can get that sorted out along with the other issues you've flaged and post the full test.

jimingham commented on packages/Python/lldbsuite/test/lang/swift/expression/exclusivity_suppression/TestExclusivitySuppression.py:
```diff
@@ -0,0 +1,145 @@
+    def test_exclusivity_suppression_on_expression_bailout(self):
+        """Test that exclusively owned values can still be accessed"""
+        self.buildAll()
+
```
Ditto for this test.

lhames replied:
As above - I'll see if I can figure out the issue I was encountering with Process continue, then post the full test.

jimingham commented on packages/Python/lldbsuite/test/lang/swift/expression/exclusivity_suppression/TestExclusivitySuppression.py:
```diff
@@ -0,0 +1,145 @@
+        report_str = "%s expected: %s got: %s" % (
+            expression, expected_result, answer)
+        self.assertTrue(answer == expected_result, report_str)
+
```
We now have 10 cut and paste copies of this function around.  We should move it to lldbutil.py and use it from there.

lhames replied:
Sounds good to me. Is that something that would be better done in a separate commit though?

jimingham replied:
Yes, that's fine.

jimingham commented on source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp:
```diff
@@ -55,6 +55,20 @@ SwiftUserExpression::SwiftUserExpression(
+  // Can we assert that the process weak-ptr is non-null here?
+  // What's the 'retry_if_fail' argument to GetSwiftLanguageRuntime?
+  // Can the runtime ever come back as null?
+  if (auto process = m_jit_process_wp.lock())
```
It should be fine to assert if no process here.  You can run lldb's expression parser in "calculator" mode without a process.  But you can only do that if the expression is IR interpretable.  This code only gets run once we decide we can't IR interpret the expression and are actually going to run it.  For that you have to have a process.

You might be debugging a C program, and somebody tries to run a Swift expression.  There is no swift loaded into your program so no swift language runtime.  The lookup is going to fail, and cache that there's no swift language runtime.  Then you load some library that has swift code.  If retry_if_null is true (the default) we will look for it again - finding it this time.  I actually can't think of a good reason to set this to false off the top of my head.

As above, the runtime can come back as null, so it's probably good to check here.  I can't see how you would be able to get to here without finding a language runtime however, so again it might be appropriate to assert.

lhames replied:
Ok. I'll add the appropriate asserts and remove those comments.

jimingham commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -3872,6 +3890,83 @@ SwiftLanguageRuntime::GetBridgedSyntheticChildProvider(ValueObject &valobj) {
+    if (bytes_read != sizeof(m_original_dynamic_exclusivity_flag_state)) {
+      if (log)
+        log->Printf("SwiftLanguageRuntime: Unable to read "
+                    "disableExclusivityChecking flag state");
```
Add the error's message to the log message (error.GetCString() will get it for you).

jimingham commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -3872,6 +3890,83 @@ SwiftLanguageRuntime::GetBridgedSyntheticChildProvider(ValueObject &valobj) {
+      if (bytes_written != sizeof(m_original_dynamic_exclusivity_flag_state)) {
+        if (log)
+          log->Printf("SwiftLanguageRuntime: Unable to set "
+                      "disableExclusivityChecking flag state");
```
Same here, put the error's message into the log output.

jimingham commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -132,6 +133,23 @@ void SwiftLanguageRuntime::SetupSwiftError() {
+  ConstString g_disableExclusivityChecking("_swift_disableExclusivityChecking");
+
+  m_dynamic_exclusivity_flag_addr = FindSymbolForSwiftObject(
+      target, g_disableExclusivityChecking, eSymbolTypeData);
```
That API (FindSymbolForSwiftObject) makes me sad.  It does nothing swift specific at all, it just looks up a symbol by name globally.  So no way does it belong in SwiftLanguageRuntime (probably Target instead).  And it doesn't actually return the symbol, it finds the symbol's address.

This function is really the same as the first half of:

Target::FindLoadAddrForNameInSymbolsAndPersistentVariables

Probably better to break that into Target::FindLoadAddrForNameInSymbols and Target::FindLoadAddrForNameInSymbolsAndPersistentVariables.  Then you can use the former for your purposes.

lhames replied:
Ok. I'll file a bug to refactor this in a follow-up.

jimingham commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -3872,6 +3890,83 @@ SwiftLanguageRuntime::GetBridgedSyntheticChildProvider(ValueObject &valobj) {
+                              &new_flag_state,
+                              sizeof(m_original_dynamic_exclusivity_flag_state),
+                              error);
+      if (bytes_written != sizeof(m_original_dynamic_exclusivity_flag_state)) {
```
Same comment as when reading.  You need to get the sizeof(bool) as above.  Then you should make a Scalar with the value of 1 (scalars don't know about bools, but a C++ bool true is an int with value 1) and you can use WriteScalarToMemory to get the byte swapping right.

jimingham commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -3872,6 +3890,83 @@ SwiftLanguageRuntime::GetBridgedSyntheticChildProvider(ValueObject &valobj) {
+      m_process->DoReadMemory(*m_dynamic_exclusivity_flag_addr,
+                              &m_original_dynamic_exclusivity_flag_state,
+                              sizeof(m_original_dynamic_exclusivity_flag_state),
+                              error);
```
You are assuming that the size of a Host bool and the size of a Target bool are the same here.  That's not guaranteed to be true.  You can get the TypeSystem from the Target (GetScratchTypeSystemForLanguage is good, that's the expression context.) then get call GetBuiltinTypeByName("bool") then get the size from the type you got back.

lhames replied:
Sounds good. I'll add the bool size query (and log error messages suggested below).

lhames replied:
Is the type system returned by GetScratchTypeSystemForLanguage going to be C++ or Swift? We need the C++ type size here (though I suspect they're the same in practice).

lhames replied:
Oh - the language is an argument to GetScratchTypeSystemForLanguage. Ignore that then. :)

jimingham commented on source/Target/SwiftLanguageRuntime.cpp:
```diff
@@ -3872,6 +3890,83 @@ SwiftLanguageRuntime::GetBridgedSyntheticChildProvider(ValueObject &valobj) {
+    if (bytes_written != sizeof(m_original_dynamic_exclusivity_flag_state)) {
+      if (log)
+        log->Printf("SwiftLanguageRuntime: Unable to reset "
+                    "disableExclusivityChecking flag state");
```
Again, add the error string here.

jimingham reviewed (changes requested):
The overall structure is fine.  See the detailed comments below.

lhames committed:
Address initial feedback from jimingham for swift-lldb PR201
(https://github.com/apple/swift-lldb/pull/201):

(1) Complete the concurrent expression test and remove the expression failure
    test until I have a better way to verify the exclusivity suppression state.

(2) Add llvm_unreachables to SwiftUserExpression WillStartExecuting /
    DidFinishExecuting. These assert that there is a process and swift runtime
    available to the expression object.

(3) Use GetBuiltinTypeByName to find the target specific size of bool, and
    ReadScalarIntegerFromMemory / WriteScalarToMemory to write the runtime's
    exclusivity suppression flag.

(4) Add error messages to logs.

lhames commented:
@swift-ci please test

jimingham commented on include/lldb/Target/SwiftLanguageRuntime.h:
```diff
@@ -485,6 +489,12 @@ class SwiftLanguageRuntime : public LanguageRuntime {
+  llvm::Optional<lldb::addr_t> m_dynamic_exclusivity_flag_addr =
+    llvm::Optional<lldb::addr_t>();
+  bool m_original_dynamic_exclusivity_flag_state = false;
+
```
Be nice to have a comment describing what these are for.

jimingham commented on source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp:
```diff
@@ -55,6 +55,29 @@ SwiftUserExpression::SwiftUserExpression(
+void SwiftUserExpression::WillStartExecuting() {
+  // Can we assert that the process weak-ptr is non-null here?
+  // What's the 'retry_if_fail' argument to GetSwiftLanguageRuntime?
+  // Can the runtime ever come back as null?
```
You don't need this comment any more, do you?

jimingham reviewed (commented):
I don't see the SwiftLangaugeRuntime.cpp changes in the latest diff here.

jimingham commented on source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp:
```diff
@@ -59,14 +59,23 @@ void SwiftUserExpression::WillStartExecuting() {
+    else
+      llvm_unreachable("Can't execute a swift expression without a runtime");
+  } else
+    llvm_unreachable("Can't execute an expression without a process");
```
An expression could cause a process to exit.  So while you can't START an expression executing without a process, it is possible that it not exist when you exit.  I am pretty sure you won't get called to here in that case, but it's probably best to just give up quietly on the way out if you don't have a process.

jimingham reviewed (commented):
Ah, okay, I found the SwiftLanguageRuntime.cpp changes by going to the latest hash directly.  That part of the diff looks fine.

lhames committed:
Address further feedback for swift-lldb PR201.

(1) Explanatory comment for exclusivity related members in
    SwiftLanguageRuntime.h.

(2) Remove assertion that a process is available from
    SwiftUserExpression::DidFinishExecuting.

(3) Remove now-redundant comment from SwiftUserExpression::WillStartExecuting.

Also removes the reference to 'frame' from the
test_exclusivity_for_concurrent_expressions test, since that variable is no
longer defined.

lhames commented:
@swift-ci please test

jimingham commented:
That looks fine.  Please remember to do the cleanups to the FindSymbolForSwiftObject and check_expression in subsequent commits.
